sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala

```diff
@@ -580,55 +580,110 @@ private[lf] object SBuiltinFun {
+            .left
+            .map(_ =>
+              IE.Dev(
+                NameOf.qualifiedNameOfCurrentFunc,
+                IE.Dev.CCTP(
+                  IE.Dev.CCTP.MalformedByteEncoding(
+                    getSText(args, 2),
+                    cause = "can not parse DER encoded public key hex string",
+                  )
+                ),
+              )
+            )
+          publicKey = extractPublicKey(derEncodedPublicKey)
+        } yield {
+          SBool(cctp.MessageSignature.verify(signature, message, publicKey))
+        }
 
-        Control.Value(SBool(cctp.MessageSignature.verify(signature, digest, publicKey)))
+        result.fold(Control.Error, Control.Value)
       } catch {
         case _: NoSuchProviderException =>
           crash("JCE Provider BouncyCastle not found")
         case _: NoSuchAlgorithmException =>
           crash("BouncyCastle provider fails to support SECP256K1")
-        case _: InvalidKeyException =>
-          // TODO: https://github.com/DACH-NY/canton-network-utilities/issues/2916: introduce structured CCTP error reporting
-          //  Control.Error(
-          //    IE.Dev(
-          //      NameOf.qualifiedNameOfCurrentFunc,
-          //      IE.Dev.CCTP(IE.Dev.CCTP.InvalidKeyError(exn.getMessage)),
-          //    )
-          //  )
-          Control.Value(SBool(false))
-        case _: InvalidKeySpecException =>
-          // TODO: https://github.com/DACH-NY/canton-network-utilities/issues/2916: introduce structured CCTP error reporting
-          //  Control.Error(
-          //    IE.Dev(
-          //      NameOf.qualifiedNameOfCurrentFunc,
-          //      IE.Dev.CCTP(IE.Dev.CCTP.InvalidKeyError(exn.getMessage)),
-          //    )
-          //  )
-          Control.Value(SBool(false))
-        case _: SignatureException =>
-          // TODO: https://github.com/DACH-NY/canton-network-utilities/issues/2916: introduce structured CCTP error reporting
-          //  Control.Error(
-          //    IE.Dev(
-          //      NameOf.qualifiedNameOfCurrentFunc,
-          //      IE.Dev.CCTP(IE.Dev.CCTP.SignatureError(exn.getMessage)),
-          //    )
-          //  )
-          Control.Value(SBool(false))
+        case exn: InvalidKeyException =>
+          Control.Error(
+            IE.Dev(
+              NameOf.qualifiedNameOfCurrentFunc,
+              IE.Dev.CCTP(IE.Dev.CCTP.MalformedKey(getSText(args, 2), exn.getMessage)),
+            )
+          )
+        case exn: InvalidKeySpecException =>
+          Control.Error(
+            IE.Dev(
+              NameOf.qualifiedNameOfCurrentFunc,
+              IE.Dev.CCTP(IE.Dev.CCTP.MalformedKey(getSText(args, 2), exn.getMessage)),
+            )
+          )
+        case exn: SignatureException =>
+          Control.Error(
+            IE.Dev(
+              NameOf.qualifiedNameOfCurrentFunc,
+              IE.Dev.CCTP(IE.Dev.CCTP.MalformedSignature(getSText(args, 0), exn.getMessage)),
+            )
+          )
```

Comment:
Which calls can produced those errors ?

Reply:
Ref: https://github.com/digital-asset/daml/blob/main/sdk/libs-scala/crypto/src/main/scala/com/daml/crypto/MessageSignaturePrototype.scala#L24-L26

Reply:
Thanks.
In this case could we limit the scope of the `try` `catch` to this unique call ?

Reply:
Would there be a benefit in doing this?

Reply:
Actually, when I look at this again, the call to `extractPublicKey` can also throw a `InvalidKeySpecException` exception.

This method can also throw a `NoSuchAlgorithmException` - but here this mean BouncyCastle is not installed and so we should crash.